db/fetcher/encoded_doc.go

```diff
@@ -178,14 +183,32 @@ func convertToInt(propertyName string, untypedValue any) (int64, error) {
 
 // @todo: Implement Encoded Document type
 type encodedDocument struct {
+	mapping *core.DocumentMapping
+	doc     *core.Doc
+
 	Key        []byte
-	Properties map[client.FieldDescription]*encProperty
+	Properties []encProperty
+
+	// tracking bitsets
+	// A value of 1 indicates a required field
+	// 0 means we we ignore the field
+	// we update the bitsets as we collect values
+	// by clearing the bit for the FieldID
+	filterSet *bitset.BitSet // filter fields
+	selectSet *bitset.BitSet // select fields
+
 }
 
 // Reset re-initializes the EncodedDocument object.
 func (encdoc *encodedDocument) Reset() {
-	encdoc.Properties = make(map[client.FieldDescription]*encProperty)
+	encdoc.Properties = make([]encProperty, 0)
 	encdoc.Key = nil
+	if encdoc.mapping != nil {
```

Comment:
suggestion: This is a bit odd without a `New` function.  Here `mapping` should never be nil no?  Yet at the same time if it ever is, the `doc` property will never be cleared by reset leaving this instance in an invalid state.

I'd suggest adding a `New(...) EncodedDocument` func that ensures `mapping` is never nil, or if for whatever reason you really dont want that, add an `else` clause that just sets `doc` to `nil`.

Reply:
Wanted a `Reset` instead of a `New` so it keeps using the same allocated/defined struct on the fetcher.

As for the `else`, the doc will already be `nil` if the `mapping` is also `nil`, and the mapping never gets updated once its set from the beginning, so its either `nil` or not for the whole thing. I can add if you you prefer, but its not a codepath that will add any meaning/safety imo.

Reply:
> Wanted a Reset instead of a New so it keeps using the same allocated/defined struct on the fetcher.

I think you misunderstand why I mentioned `New`, by declaring a constructor you are declaring a restricted initial state.  There you can enforce `mapping` is not nil and reset loses a dead code branch (the implicit else), as well as possibility of the invalid `doc` state that I mentioned. `New` can even call `Reset` internally for a stronger initial state guarantee/description if you want.